### ISSUE TITLE
Support for null in list items

### ISSUE THREAD

**Issue by renatocron** [prisma/prisma#15886]
## Problem

Some times, you want to have positional data saved onto arrays, and some of the slots are empty (NULL)
Currently, prisma does not support this

## Suggested solution

I think that's a bit edge case, so I think maybe on the schema you can Argument the column with somehting like @AcceptNullsItems() in order to make the types of each item as (T | null)[] instead of T[]

## Alternatives

Make the (T | null) the default behavior, as it is in the database (at least postgres)

**Comment by aprilmintacpineda** [comment]
Workaround:

```
type OptionalList {
  list String[]
}

model MyModel {
  options OptionalList?
}
```

But you have to make sure to programmatically correct your code.

**Comment by renatocron** [comment]
Hello, @aprilmintacpineda,

I don't think I understand how your workaround would function the same way. I believe the workaround using types would look something like this:

```
type OptionalBooleanValueType {
  value Bool?
}

model MyModel {
  initialized_array OptionalBooleanValueType[]
}
```
So it can store an array of elements (required), where the individual elements (not the entire array itself) can be null.

However, at this point, improved database modeling would be preferable than dealing with this peculiar array of enum with a single column. Regardless, I initiated this thread to log the issue, as PostgreSQL supports it, but Prisma currently does not have a way to represent it.

**If anyone here is from the core team, please do not expend too much time on this!** instead, focus on enhancing missing features, such as partial indexes!

Adding this feature could complicate Prisma's implementation, potentially introducing new bugs on the schema syntax, and I can't event think about the complexities on the rust engine to deal with that.

**Comment by aprilmintacpineda** [comment]
@renatocron this issue was originally at https://github.com/prisma/prisma/issues/847 which talked about optional list:

```prisma
{
  value String[]?
}
```

**Comment by renatocron** [comment]
Yes, it still missing, I originally commented there but when I realied it was different issued, I opened this issue (https://github.com/prisma/prisma/issues/847 was already closed as 'not fixed' I think back then, as commented by https://github.com/prisma/prisma/issues/847#issuecomment-683666269) and edited to metion here (it was already closed)
now there's another https://github.com/prisma/prisma/issues/14390 that is about the issue where your workaround would indeed work

**Comment by aprilmintacpineda** [comment]
Ah I see, I think https://github.com/prisma/prisma/issues/847 and https://github.com/prisma/prisma/issues/14390 are more related than https://github.com/prisma/prisma/issues/847 and this issue.

I will post my workaround there, should I delete that comment here though? Just to keep this issue clean as well.

**Comment by Jacse** [comment]
As more and more people start storing vectors (embeddings) in their dbs to support AI use cases this is also an issue. I want to ensure that all vectors are the correct format and a fixed length (or no vector at all), ideally that would be typed as null or vector. Empty vector/length 0 vs actual vector makes it much harder to query (I can't just filter all rows that have a vector I have to check their length at query time).

**Comment by renatocron** [comment]
Just updating, my comments about not wasting time wre about the null inside the list. 

I think having a optionally null array should be supported, I don't get why the behavior was different from enum or other field types, this can be frustrating to more users as well

**Comment by blakew** [comment]
Hi, any update here?

**Comment by henriqueArrazao** [comment]
The work around I'm using is setting `@default([])` in the schema. Although the null makes much more sense. Any update??

**Comment by PrathamSoni** [comment]
Bump here! would be very helpful for us...

**Comment by LemonFace0309** [comment]
Hey, we're running into a critical issue with Prisma now where an aggregated materialized view we use requires having null values in the list to ensure each agg list column in our table is of equal length. We weren't able to find any suitable replacement and it's currently in conflict with our development. Would appreciate updates here

**Comment by renatocron** [comment]
> The work around I'm using is setting `@default([])` in the schema. Although the null makes much more sense. Any update??

This *is not this* issue, tho, this was about having null elements, not the field itself.

I'm updating this so we can see how this is going on Prisma 6 with new rust-less engine / Prisma 7

**Comment by KaKi87** [comment]
Hi,

Any news on this please ?

Thanks

---

@pantharshit00

> you want to return `[]` instead of `null` if there are no items in the lists

No. I want to write `if(list) for(item of list){} else doSomething()`, not `if(list.length) for(item of list){} else doSomething()`.

**Comment by tom-plaa** [comment]
Also asking, this has forced us to use really ugly workarounds, like using sentinel values of -1e30...
